### ISSUE TITLE
Subsequent calls to tf.data.Dataset.map() with seeded random operation give same random sequences.

### ISSUE THREAD

**Issue by johannabar** [tensorflow/tensorflow#35090]
**System information**
- Have I written custom code: yes
- OS Platform and Distribution: CentOS 7
- TensorFlow installed from (source or binary): binary
- TensorFlow version: 1.14.0
- Python version: 2.7.17
- CUDA/cuDNN version: 10.0

**Describe the current behavior**
In eager mode when setting a random seed, two calls to the same mapping function in tf.data.Dataset.map() would result in two times the same generated random values.

**Describe the expected behavior**
As I haven't explicitly closed or restarted a session, I would expect both calls to output different generated random values. Is tf.data.Dataset.map() creating its own session and running the mapping function subgraph in it? If so, is there a way to force the map() to take the current session as input?

**Code to reproduce the issue**
```
import numpy as np  
import tensorflow as tf
tf.compat.v1.enable_eager_execution()

SEED = 88
tf.compat.v1.random.set_random_seed(SEED)

ds_train = tf.data.Dataset.range(0, 4)
ds_val = tf.data.Dataset.range(0, 4)
ds_train = ds_train.map(
    lambda x: tf.random.uniform([1], 0.2, 5.0)
)
ds_val = ds_val.map(
    lambda x: tf.random.uniform([1], 0.2, 5.0) 
)
for el in ds_train:
    print(el.numpy())
    # --> [0.44027787], [1.7892183], [2.8793733], [3.3438706]
for el in ds_val:
    print(el.numpy())
    # why the same here? --> [0.44027787], [1.7892183], [2.8793733], [3.3438706]
```

**Comment by gadagashwini** [comment]
I could replicate the issue on colab with Tf 1.14 and 1.15.
Please find the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/7840922356bfca9f85923b8271f3aa8c/untitled305.ipynb). Thanks!

**Comment by ymodak** [comment]
```tf.random.uniform``` has a ```seed```argument which is same in both calls. Therefore you see same numbers when you print.
If the graph-level seed is set, but the operation seed is not: The system deterministically picks an operation seed in conjunction with the graph-level seed so that it gets a unique random sequence.
See https://www.tensorflow.org/versions/r1.15/api_docs/python/tf/random/set_random_seed

**Comment by johannabar** [comment]
Hi @ymodak , thank you for your answer!
I understand the double-seed system but in the case where you have the same random operation with the same seed, you should still have a sequence of different random numbers within the same session. This sequence should repeat only if you compute it in a new session. For instance
```
import tensorflow as tf
tf.compat.v1.enable_eager_execution()
SEED = 88
tf.compat.v1.random.set_random_seed(SEED)
rand1 = tf.random.uniform([1], 0.2, 5.0, seed=SEED) #-->[2.7339704]
rand2 = tf.random.uniform([1], 0.2, 5.0, seed=SEED) #-->[1.4490409] different
# Same when only graph-level seed is set
rand1 = tf.random.uniform([1], 0.2, 5.0) #-->[2.9647074]
rand2 = tf.random.uniform([1], 0.2, 5.0) #-->[2.7952404] different
```
Which is why I would expect both random operation in the map() function to be a continuation of a sequence of different random numbers as we are still in the same session.

**Comment by wangpengmit** [comment]
Hi @johannabar, there are no longer sessions when eager mode is enabled. Every time `Dataset.map` is called, it creates a `tf.function` for the mapping function. The `random.uniform` in the mapping function gets its graph-level seed from the `tf.function`'s default graph's seed (https://github.com/tensorflow/tensorflow/blob/1cf0898dd4331baf93fe77205550f2c2e6c90ee5/tensorflow/python/framework/random_seed.py#L62). I expected this seed to be `None`, but apparently it's not (maybe this is a side effect of `set_random_seed`), in which case the two `random.uniform`s start from the same seed and the same state (because both are in a fresh new `tf.function`). BTW you can check out the new RNG (https://www.tensorflow.org/api_docs/python/tf/random/experimental/Generator) which has much better seeding control.

**Comment by johannabar** [comment]
Hi @wangpengmit, thank you for the explanation!
I am closing this issue as what I noticed is correct behaviour.

Though I think it's a little tricky that a map call to the same mapping function will produce the same random numbers. For instance if you have a map function doing random cropping on images and you call it on both your training and validation data separately then you will get the SAME cropping positions for both training and validation data, is that right?

**Comment by wangpengmit** [comment]
Yes. And I confirm that `set_random_seed` will cause an RNG in `tf.function` to start from the same global seed (because of https://github.com/tensorflow/tensorflow/blob/1cf0898dd4331baf93fe77205550f2c2e6c90ee5/tensorflow/python/eager/context.py#L427 and https://github.com/tensorflow/tensorflow/blob/1cf0898dd4331baf93fe77205550f2c2e6c90ee5/tensorflow/python/framework/func_graph.py#L229). Possible solutions for your code are removing `set_random_seed` call or calling it before each `Dataset.map` with different seeds.

**Comment by winthropharvey** [comment]
> Yes. And I confirm that `set_random_seed` will cause an RNG in `tf.function` to start from the same global seed (because of
> 
> https://github.com/tensorflow/tensorflow/blob/1cf0898dd4331baf93fe77205550f2c2e6c90ee5/tensorflow/python/eager/context.py#L427
> and
> 
> https://github.com/tensorflow/tensorflow/blob/1cf0898dd4331baf93fe77205550f2c2e6c90ee5/tensorflow/python/framework/func_graph.py#L229
> ). Possible solutions for your code are removing `set_random_seed` call or calling it before each `Dataset.map` with different seeds.

I am having a very similar issue.  I can confirm that removing the graph level random seed indeed makes things random across batches as desired.  However, I still want the ability to seed in order to have a consistent validation set when doing some random data augmentation.  

As regards your second suggestion, to call a seed before each map call in a defined order - this doesn't seem to be possible.  To do this, you need to have some external state being tracked that is then modified by the map function. Sadly, I don't seem to be able to find any way to persist an external state through map calls to update seed in a defined order unless I use the py_function decorator, which slows down computation a lot and can produce other issues.  Do you have any suggestions?  

Specifically, what I'm trying to do is 

image_crop = tfa.image.random_cutout(image_crop, [20,20], 0, seed)

within a map function on a dataset.  I need seed to progress on every call of this function.  If I could make seed a global variable, I can do this with a simple seed += 1.  In fact, this works perfectly fine if the map function is contained in a py_function.  However, when it is not, you can not alter any python global variable inside a tf.function.  I can't find any solution that allows me to track an external state so I know how many times map function has been called unless I use py_function.  Let me know if you have any solutions/ideas.  I can always use py_function if I have to, but I'd prefer to avoid the slowdown.

**Comment by philippemiron** [comment]
Hi, I seem to be having the same issues. I'm trying to augment an image and a mask and obviously used the same transformation.

I started from this sample example from the [documentation](https://www.tensorflow.org/tutorials/images/segmentation#download_the_oxford-iiit_pets_dataset).

```
class Augment(tf.keras.layers.Layer):
  def __init__(self, seed=42):
    super().__init__()
    # both use the same seed, so they'll make the same random changes.
    self.augment_inputs = tf.keras.layers.RandomFlip(mode="horizontal", seed=seed)
    self.augment_labels = tf.keras.layers.RandomFlip(mode="horizontal", seed=seed)

  def call(self, inputs, labels):
    inputs = self.augment_inputs(inputs)
    labels = self.augment_labels(labels)
    return inputs, labels
```

If I run the generators multiple times, flips are inconsistent between the images and masks.

```
# Create a tf.datasets
ds = tf.data.Dataset.from_tensor_slices((fields, masks))

# plot to verify
for f, m in ds.map(Augment()).take(5):
    show_image(f.numpy(), m.numpy())
```

I tried instead to set the `seed` inside the class instead of fixing it per layer as suggested in this thread,`tf.random.set_seed(seed)`, but the issue is still present.

**Comment by jaroslawjanas** [comment]
> Hi, I seem to be having the same issues (on TF 2.10.0). I'm trying to create an augmentations pipeline to process images and masks.
> 
> I started from this sample example from the [documentation](https://www.tensorflow.org/tutorials/images/segmentation#download_the_oxford-iiit_pets_dataset).
> 
> ```
> class Augment(tf.keras.layers.Layer):
>   def __init__(self, seed=42):
>     super().__init__()
>     # both use the same seed, so they'll make the same random changes.
>     self.augment_inputs = tf.keras.layers.RandomFlip(mode="horizontal", seed=seed)
>     self.augment_labels = tf.keras.layers.RandomFlip(mode="horizontal", seed=seed)
> 
>   def call(self, inputs, labels):
>     inputs = self.augment_inputs(inputs)
>     labels = self.augment_labels(labels)
>     return inputs, labels
> ```
> 
> If I run the generators multiple times, flips are inconsistent between the images and masks.
> 
> ```
> # Create a tf.datasets
> ds = tf.data.Dataset.from_tensor_slices((fields, masks))
> 
> # plot to verify
> for f, m in ds.map(Augment()).take(5):
>     show_image(f.numpy(), m.numpy())
> ```
> 
> I tried instead to set the `seed` inside the class instead of fixing it per layer as suggested in this thread,`tf.random.set_seed(seed)`, but the issue is still present.
> 
> Notes: I created a full example and shared on [StackOverflow](https://stackoverflow.com/questions/74764831/preprocessing-layers-with-seed-not-producing-the-same-data-augmentation-for-imag?noredirect=1#comment131954709_74764831).

The fact that the documentation suggests [this](https://www.tensorflow.org/tutorials/images/segmentation) but it doesn't work as intended is very concerning.
The seeds seem to be diverging. In my case the **mask** is sometimes flipped differently than the **input image**.
Even if I am doing something wrong or there is a better way to do this the documentation should be updated.

The best workaround I found on StackOverflow was to merge the **input image** and **mask** by channel axis and then perform augmentation on the combined image.

**Comment by vscv** [comment]
In TensorFlow v2.13.0, RandomFlip and RandomRotation produce augmented images and masks with identical enhancements when using the same seed but tf.keras.layers.RandomTranslation somehow does not work like that.
